### DISCUSSION TITLE
Signing, Renewal and CRL error in 2.5.0

### DISCUSSION THREAD

**Discussion by matthiasradde** [chris2511/xca#468] [Q&A] [answered]
I've updated XCA from 2.4.0 to 2.5.0 and opened the xca-database-file.
Tried to extend an existing certificate (by one year) which will expire within a few days.
Resulted in following errormessage
```
Der folgende Fehler ist aufgetreten:
(8pki_x509[]:corei5-10_4)
error:0300009C:digital envelope routines::unsupported algorithm
error:068C0100:asn1 encoding routines::malloc failure
error:068C0100:asn1 encoding routines::malloc failure

(C:\Users\chris\build\xca-2.5.0\lib\pki_x509.cpp:582)
```
Uninstalled 2.5.0 and installed 2.4.0. Now extending the same certificate was successful.

Certificate to be extended was signed with sha512WithRSAEncryption (OID 1.2.840.113549.1.1.13) and used with its 4096-bit-key.

What did I do wrong? Or what additional information is needed to reproduce this issue?

**Comment by chris2511** [comment]
I could not reproduce here on Windows 10 with xca-2.5.0 and sha512WithRSAEncryption with 4096bit RSA key
Do you use the installed version or the portable?
What library versions does the "About" dialog show?

**Comment by lwt-pressy** [comment]
similar problem here with generating a crl. Also with sha512WithRSAEncryption and 4096-bit-key

> (7pki_crl[]:TERTA-Zertifizierungs CA 2015)
> error:0300009C:digital envelope routines::unsupported algorithm
> error:068C0100:asn1 encoding routines::malloc failure
> error:068C0100:asn1 encoding routines::malloc failure
> 
> (C:\Users\chris\build\xca-2.5.0\lib\pki_crl.cpp:250)


working Version:

----------------------------------------
> Version: 2.4.0
> ECC With RFC 5639 Brainpool curves
> OpenSSL 1.1.1k 25 Mar 2021
> QT version: 5.12.0

not working:
----------------------------------------

> Copyright 2001 - 2023 by Christian Hohnstädt 
> Version: 2.5.0
> OpenSSL 3.1.2 1 Aug 2023
> QT version: 6.5.2

I'm not sure if it is relevant, but I select always typical install.

**Comment by matthiasradde** [comment]
re-checked the issue - always used full installation on Windows 10 - not the portable version

XCA - working

> Copyright 2001 - 2021 by Christian Hohnstädt 
> Version: 2.4.0
> ECC With RFC 5639 Brainpool curves
> OpenSSL 1.1.1k 25 Mar 2021
> QT version: 5.12.0
> https://hohnstaedt.de/xca
> Entropy strength: 110
> 
> Installation path:
> C:\Program Files\xca
> User settings path:
> C:\Users\Matthias\AppData\Roaming\xca
> Working directory:
> C:\Users\Matthias\Desktop\


XCA - non working

> Copyright 2001 - 2023 by Christian Hohnstädt 
> Version: 2.5.0
> OpenSSL 3.1.2 1 Aug 2023
> QT version: 6.5.2
> https://hohnstaedt.de/xca
> Entropy strength: 40
> 
> Installation path:
> C:\Users\Matthias\AppData\Roaming\xca
> User settings path:
> C:\Users\Matthias\AppData\Roaming\xca
> Working directory:
> C:\Users\Matthias\Desktop\


checked (re-)creating a CRL

> Der folgende Fehler ist aufgetreten:
> (7pki_crl[]:radde-ca-server)
> error:0300009C:digital envelope routines::unsupported algorithm
> error:068C0100:asn1 encoding routines::malloc failure
> error:068C0100:asn1 encoding routines::malloc failure
> 
> (C:\Users\chris\build\xca-2.5.0\lib\pki_crl.cpp:250)

extend a certificate

> Der folgende Fehler ist aufgetreten:
> (8pki_x509[]:corei5-10_4)
> error:0300009C:digital envelope routines::unsupported algorithm
> error:068C0100:asn1 encoding routines::malloc failure
> error:068C0100:asn1 encoding routines::malloc failure
> 
> (C:\Users\chris\build\xca-2.5.0\lib\pki_x509.cpp:582)

**Comment by chris2511** [comment]
According to #461 and #410 this issue does not depend on the operating system ....

**Comment by SemoTech** [comment]
Hi @chris2511 thats very strange as 2.4.0 works fine to create.
I wonder if there is something specific about my setup or that of @matthiasradde 
Interestingly it is failing on Windows for him and on [Mac OS 12.7 for me](https://github.com/chris2511/xca/issues/461)...
The failure occurs both when extending as well as creating new certs with v2.5.0

**Comment by SemoTech** [comment]
@chris2511 can you maybe add a detailed debugging option into a beta build, link it here, and then we can both submit the logs to you?

**Comment by amette** [comment]
I have the same error on 2.4.0 when trying to create a new certificate. Key generation seems to have been successful.

**Comment by chris2511** [comment]
> I have the same error on 2.4.0 when trying to create a new certificate. Key generation seems to have been successful.

You mean 2.5.0 ? (Just to be sure, because all other observations say: works in 2.4.0, fails in 2.5.0)

**Reply by amette** [reply]
No, I meant 2.4.0. I just tried changing the password, which didn't work and yielded the following error:

The following error occurred:
(7pki_evp[1]:CCI CA)
error:1D8C0102:ENCODER routines::passed a null parameter

(pki_evp.cpp:573)

**Reply by amette** [reply]
I was able to change the password with XCA 2.4.0 on a Debian 12 machine. Then I copied the database back to my Gentoo machine and was able to create certificates with XCA 2.4.0 on Gentoo.

Edit:  XCA's About page says "Compile time: OpenSSL 3.0.3 3 May 2022 / Run time: OpenSSL 3.0.9 30 May 2023" on Debian and "OpenSSL 3.0.10 1 Aug 2023" on Gentoo. Just in case that might be related.

**Reply by 6XGate** [reply]
There seems to be a similar issue on Ubuntu 22.04. Using XCA 2.4.0 to fix the legacy DB keys results in the null parameter error on some databases but not all. Compile time version: OpenSSL 3.0.3 3 May 2022, runtime version: OpenSSL 3.0.11 19 Sep 2023.

**Comment by chris2511** [comment, marked as answer]
There is a solution (I think):
The common cause is the age. Your database was created before 2.0.0 and I dropped support for the old database password encryption with XCA 2.5.0. However, the **keys** were not re-encrypted with the new PKCS#8 format during upgrade to the SQL database scheme, which did not matter, because XCA until 2.4.0 was still able to read them.

The malfunctioning **key** should say "Legacy database" in the "Context-menu"-> Properties->Source.
The encryption scheme needs an update. XCA-2.4.0 can do this:

 -  Select "Change database Password" with XCA 2.4.0 
 -  If you have keys with a private password, select "Reset password" and then "Change password" again for those keys.

In both cases, the old and new password may be the same.
Afterwards XCA 2.5.0 should work.

**Reply by matthiasradde** [reply]
Thank you! I can confirm that - after changing the database's password with 2.4.0 - both issues are now working with 2.5.0.
Only thing ist that the affected key's source said "Erstellt" (="created") instead of "Alte Datenbank" (="Legacy database"). But important thing is: it is working now 👍

**Reply by lwt-pressy** [reply]
updating the password, did also work for me. I could create the new crls without problems, thx.

**Reply by 6XGate** [reply]
Looks like this solution may be unusable for some using Linux when the older XCA version is no longer available for their system.

**Reply by Bgs4269** [reply]
You can downgrade using another package. e.g. fc39 comes with 2.5.0, but you can get a 2.4.0 repackage/rebuild here (I used this personally): https://rpmfind.net/linux/RPM/fedora/39/x86_64/x/xca-2.4.0-5.fc39.x86_64.html
Look for one specific to your distro.

**Reply by 6XGate** [reply]
Discovered a few days later Distrobox can be used to run older XCA builds from other distros as well.

**Reply by gwtad** [reply]
After going through this process, the keys still show source "Legacy Database". Is this expected? Is there a way to upgrade those keys?

**Reply by MarcinOrlowski** [reply]
Hint: database file is cross platform. Once you "fix" your file i.e. using XCA v24.0 Portable on Windows you can then copy back and continue using it i.e. on Linux.

**Reply by claasz** [reply]
I successfully managed to install the 2.4 version on Ubuntu 24.04 by downloading https://github.com/chris2511/xca/releases/download/RELEASE.2.4.0/xca-2.4.0.tar.gz and following the instructions in `INSTALL.docker`.

**Comment by SemoTech** [comment]
**@chris2511 You're a genius!  It worked!**  Resetting the password of the DB (using same pass) in 2.4.0 and opening in 2.5.0 was the fix!

If I may be so bold, can we have just a small aesthetic fix in the next version, and take advantage of wasted space on the bottom of the app to show longer DB paths/name?  
Maybe even move the "Search" box further to the right and make it a bit smaller to make even more room for longer paths?

![image](https://github.com/chris2511/xca/assets/1452100/20a0c5f0-0edc-448f-a1da-57b0df3f5a7e)

Thanks in advance!
